# Project 1: SAT & ACT Analysis

The first markdown cell in a notebook is a great place to provide an overview of your entire project. You will likely want to at least state your

## Problem Statement

By analyssis the data, I want to find the relationship between:
1. SAT and ACT score and attend rate. 
2. Visualize to povid that increase the attend rate help to increase the score. 

## Executive Summary

If you want to, it's great to use relative links to direct your audience to various sections of a notebook. **HERE'S A DEMONSTRATION WITH THE CURRENT SECTION HEADERS**:

### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [233]:
#Imports packages
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [234]:
# Read the SAT & ACT
act_2017 = pd.read_csv('../data/act_2017.csv')
sat_2017 = pd.read_csv('../data/sat_2017.csv')

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [235]:
act_2017.head()

State Participation  English  Math  Reading  Science Composite
0  National           60%     20.3  20.7     21.4     21.0      21.0
1   Alabama          100%     18.9  18.4     19.7     19.4      19.2
2    Alaska           65%     18.7  19.8     20.4     19.9      19.8
3   Arizona           62%     18.6  19.8     20.1     19.8      19.7
4  Arkansas          100%     18.9  19.0     19.7     19.5      19.4

In [236]:
sat_2017.head()

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            5%                                 593   572   1165
1      Alaska           38%                                 547   533   1080
2     Arizona           30%                                 563   553   1116
3    Arkansas            3%                                 614   594   1208
4  California           53%                                 531   524   1055

In [237]:
sat_2017.describe()

Evidence-Based Reading and Writing        Math        Total
count                           51.000000   51.000000    51.000000
mean                           569.117647  547.627451  1126.098039
std                             45.666901   84.909119    92.494812
min                            482.000000   52.000000   950.000000
25%                            533.500000  522.000000  1055.500000
50%                            559.000000  548.000000  1107.000000
75%                            613.000000  599.000000  1212.000000
max                            644.000000  651.000000  1295.000000

In [238]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null object
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [239]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
State            52 non-null object
Participation    52 non-null object
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null object
dtypes: float64(4), object(3)
memory usage: 2.9+ KB


In [240]:
act_2017.describe()

English       Math    Reading    Science
count  52.000000  52.000000  52.000000  52.000000
mean   20.919231  21.173077  22.001923  21.040385
std     2.332132   1.963602   2.048672   3.151113
min    16.300000  18.000000  18.100000   2.300000
25%    19.000000  19.400000  20.475000  19.900000
50%    20.550000  20.900000  21.700000  21.150000
75%    23.300000  23.100000  24.125000  22.525000
max    25.500000  25.300000  26.000000  24.900000

In [241]:
sat_2017.shape

(51, 5)

In [242]:
act_2017.shape

(52, 7)

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

1) SAT and ACT are the test that prove for all the student in united states. The dataframe above shows the participation rate, score's for all sections and the total score.  
2) ACT score has national average, but SAT score hasn't.
3) Some of the states take ACT score as primary, the others are more focus on SAT, so the participation rate is different. 

#### 4a. Does the data look complete? 

Answer:
Yes. The table doesn't have missing data, although it has some errors. 

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

In [243]:
act_2017.max()

State            Wyoming
Participation        98%
English             25.5
Math                25.3
Reading               26
Science             24.9
Composite           25.5
dtype: object

A) The ACT Composite score of Wyoming state is 20.2x instead of 20.2.
B)
SAT
Min
Participation                              2%
Evidence-Based Reading and Writing        482
Math                                       52
Total                                     950

Max
Participation                            100%
Evidence-Based Reading and Writing        644
Math                                      651
Total                                    1295


ACT
Min
Participation         8%
English             16.3
Math                  18
Reading             18.1
Science              2.3
Composite           17.8

Max
Participation       100%
English             25.5
Math                25.3
Reading               26
Science             24.9
Composite           25.5

#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

In [244]:
act_2017 = act_2017.replace("20.2x", "20.2")
act_2017.tail()

State Participation  English  Math  Reading  Science Composite
47       Virginia           29%     23.5  23.3     24.6     23.5      23.8
48     Washington           29%     20.9  21.9     22.1     22.0      21.9
49  West Virginia           69%     20.0  19.4     21.2     20.5      20.4
50      Wisconsin          100%     19.7  20.4     20.6     20.9      20.5
51        Wyoming          100%     19.4  19.8     20.8     20.6      20.2

#### 5. What are your data types? 
Display the data types of each feature. 

In [245]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
State            52 non-null object
Participation    52 non-null object
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null object
dtypes: float64(4), object(3)
memory usage: 2.9+ KB


In [246]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null object
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

Answer:  
A) the participation rate is object, the ACT dataframe is int
B)
1) The participation rate should be covert from object to float
2) Covert act dataframe from int to float
3) Add a national avgerage for SAT

#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

In [247]:
act_2017.columns

Index(['State', 'Participation', 'English', 'Math', 'Reading', 'Science',
       'Composite'],
      dtype='object')

### Three different way to use inside the def
- for loop : calling the cell in column and convert to float
- .astype(float) : function we can use to convert 'Series' in Pandas to float
- .map() : use function to do-somthing in 'Series'
    - ex: use lambda cell: function(cell)

### Method 1: for loop 
```def as_float(dataframe, column_name):
        for cell in dataframe[column_name]:
            float(cell)
        return dataframe```
    
### Method 2: .astpye()  

```def as_float(dataframe, column_name):
        dataframe[column_name].astype(float)
        return dataframe```

In [248]:
act_2017.head()

State Participation  English  Math  Reading  Science Composite
0  National           60%     20.3  20.7     21.4     21.0      21.0
1   Alabama          100%     18.9  18.4     19.7     19.4      19.2
2    Alaska           65%     18.7  19.8     20.4     19.9      19.8
3   Arizona           62%     18.6  19.8     20.1     19.8      19.7
4  Arkansas          100%     18.9  19.0     19.7     19.5      19.4

In [249]:
act_2017['Participation'] = act_2017['Participation'].map(lambda part : part.replace("%" , ""))

In [250]:
# Method 3: .map() use function to do-somthing in 'Series'   
#           + lambda: work for individaul cell, ex: cell: function(cell)
#           + float(): work for individaul cell, in this case, following the lambda
def as_float(dataframe, column_name):    
    dataframe[column_name] = dataframe[column_name].map(lambda part : float(part))
    return dataframe


In [251]:
# Combine everything in a line
# act_2017['Participation']= act_2017['Participation'].map(lambda part : float(part.replace("%" , ""))/100)


In [252]:
act_2017_new = as_float(act_2017, "Participation")

In [253]:
act_2017_new['Participation']= act_2017_new['Participation']/100
act_2017_new.head()

State  Participation  English  Math  Reading  Science Composite
0  National           0.60     20.3  20.7     21.4     21.0      21.0
1   Alabama           1.00     18.9  18.4     19.7     19.4      19.2
2    Alaska           0.65     18.7  19.8     20.4     19.9      19.8
3   Arizona           0.62     18.6  19.8     20.1     19.8      19.7
4  Arkansas           1.00     18.9  19.0     19.7     19.5      19.4

In [254]:
sat_2017['Participation']= sat_2017['Participation'].map(lambda part : float(part.replace("%" , ""))/100)
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null float64
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


- Fix any individual values preventing other columns from being the appropriate type.

In [255]:
act_2017_new = as_float(act_2017, "Composite")

In [256]:
sat_2017_new = as_float(sat_2017, "Evidence-Based Reading and Writing")
sat_2017_new = as_float(sat_2017, "Math")
sat_2017_new = as_float(sat_2017, "Total")

- Finish your data modifications by making sure the columns are now typed appropriately.

In [257]:
act_2017_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
State            52 non-null object
Participation    52 non-null float64
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null float64
dtypes: float64(6), object(1)
memory usage: 2.9+ KB


In [258]:
sat_2017_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null float64
Evidence-Based Reading and Writing    51 non-null float64
Math                                  51 non-null float64
Total                                 51 non-null float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


- Display the data types again to confirm they are correct.

In [259]:
act_2017_new.head()

State  Participation  English  Math  Reading  Science  Composite
0  National           0.60     20.3  20.7     21.4     21.0       21.0
1   Alabama           1.00     18.9  18.4     19.7     19.4       19.2
2    Alaska           0.65     18.7  19.8     20.4     19.9       19.8
3   Arizona           0.62     18.6  19.8     20.1     19.8       19.7
4  Arkansas           1.00     18.9  19.0     19.7     19.5       19.4

In [260]:
sat_2017_new.head()

State  Participation  Evidence-Based Reading and Writing   Math  \
0     Alabama           0.05                               593.0  572.0   
1      Alaska           0.38                               547.0  533.0   
2     Arizona           0.30                               563.0  553.0   
3    Arkansas           0.03                               614.0  594.0   
4  California           0.53                               531.0  524.0   

    Total  
0  1165.0  
1  1080.0  
2  1116.0  
3  1208.0  
4  1055.0

#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [261]:
act_2017_new_col = {'State':'state' ,
                     'Participation':'participation_act17',
                     'English':'english_act17',
                     'Math':'math_act17',
                     'Reading':'reading_act17',
                     'Science':'science_act17',
                     'Composite':'total_act17'} 
act_2017_new.rename(columns = act_2017_new_col, inplace = True)
act_2017_new.head()

state  participation_act17  english_act17  math_act17  reading_act17  \
0  National                 0.60           20.3        20.7           21.4   
1   Alabama                 1.00           18.9        18.4           19.7   
2    Alaska                 0.65           18.7        19.8           20.4   
3   Arizona                 0.62           18.6        19.8           20.1   
4  Arkansas                 1.00           18.9        19.0           19.7   

   science_act17  total_act17  
0           21.0         21.0  
1           19.4         19.2  
2           19.9         19.8  
3           19.8         19.7  
4           19.5         19.4

In [262]:
sat_2017.columns

Index(['State', 'Participation', 'Evidence-Based Reading and Writing', 'Math',
       'Total'],
      dtype='object')

In [263]:
sat_2017_new = sat_2017
sat_2017_new_col = {'State':'state' ,
                     'Participation':'participation_sat17',
                     'Evidence-Based Reading and Writing':'reading_writing_sat17',
                     'Math':'math_sat17',
                     'Total':'total_sat17'} 
sat_2017_new.rename(columns = sat_2017_new_col, inplace = True)
sat_2017_new.head()

state  participation_sat17  reading_writing_sat17  math_sat17  \
0     Alabama                 0.05                  593.0       572.0   
1      Alaska                 0.38                  547.0       533.0   
2     Arizona                 0.30                  563.0       553.0   
3    Arkansas                 0.03                  614.0       594.0   
4  California                 0.53                  531.0       524.0   

   total_sat17  
0       1165.0  
1       1080.0  
2       1116.0  
3       1208.0  
4       1055.0

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

|Feature              |Type    |Dataset |Description|
|---------------------|--------|------- |-----------|
|participation_sat17  |float   |SAT 2017|The probability of participation rate of SAT Test Score for each state.|
|reading_writing_sat17|float   |SAT 2017|Evidence-Based Reading and Writing Section Score of 2017 SAT Test for each state.|  
|math_sat17           |float   |SAT 2017|Math Section Score of 2017 SAT Test for each state.|
|total_sat17          |float   |SAT 2017|Total Score of 2017 SAT Test for each state.|
|xxx_sat18            |---     |SAT 2018|Infomation of SAT Test hold in 2018 for each state.|
|participation_act17  |float   |ACT 2017|The probability of participation rate of ACT Test Score for each state.|
|english_act17        |float   |ACT 2017|English Section Score of 2017 ACT Test for each state.|
|math_act17           |float   |ACT 2017|Math Section Score of 2017 ACT Test for each state.|
|reading_act17        |float   |ACT 2017|Reading Section Score of 2017 ACT Test for each state.|
|science_act17        |float   |ACT 2017|Science Section Score of 2017 ACT Test for each state.|
|total_act17          |float   |ACT 2017|Total Score of 2017 ACT Test for each state.|

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [264]:
sat_2017_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
state                    51 non-null object
participation_sat17      51 non-null float64
reading_writing_sat17    51 non-null float64
math_sat17               51 non-null float64
total_sat17              51 non-null float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


In [265]:
act_2017_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
state                  52 non-null object
participation_act17    52 non-null float64
english_act17          52 non-null float64
math_act17             52 non-null float64
reading_act17          52 non-null float64
science_act17          52 non-null float64
total_act17            52 non-null float64
dtypes: float64(6), object(1)
memory usage: 2.9+ KB


In [266]:
act_2017_new = act_2017_new.drop([0])

In [267]:
act_2017_new.head()

state  participation_act17  english_act17  math_act17  reading_act17  \
1     Alabama                 1.00           18.9        18.4           19.7   
2      Alaska                 0.65           18.7        19.8           20.4   
3     Arizona                 0.62           18.6        19.8           20.1   
4    Arkansas                 1.00           18.9        19.0           19.7   
5  California                 0.31           22.5        22.7           23.1   

   science_act17  total_act17  
1           19.4         19.2  
2           19.9         19.8  
3           19.8         19.7  
4           19.5         19.4  
5           22.2         22.8

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [268]:
#DataFrame.merge(self, right, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, sort=False, suffixes=('_x', '_y'), copy=True, indicator=False, validate=None
combined_2017 = pd.DataFrame.merge(self = sat_2017_new, right= act_2017_new, how = 'inner')
combined_2017.head()


state  participation_sat17  reading_writing_sat17  math_sat17  \
0     Alabama                 0.05                  593.0       572.0   
1      Alaska                 0.38                  547.0       533.0   
2     Arizona                 0.30                  563.0       553.0   
3    Arkansas                 0.03                  614.0       594.0   
4  California                 0.53                  531.0       524.0   

   total_sat17  participation_act17  english_act17  math_act17  reading_act17  \
0       1165.0                 1.00           18.9        18.4           19.7   
1       1080.0                 0.65           18.7        19.8           20.4   
2       1116.0                 0.62           18.6        19.8           20.1   
3       1208.0                 1.00           18.9        19.0           19.7   
4       1055.0                 0.31           22.5        22.7           23.1   

   science_act17  total_act17  
0           19.4         19.2  
1           19.9         19.8  
2           19.8         19.7  
3           19.5         19.4  
4           22.2         22.8

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [269]:
#df = pd.DataFrame(dict) saving the dataframe df.to_csv('file1.csv') 
combined_2017.to_csv("combined_2017.csv")

## 2018 Data Import and Cleaning

Links to the 2018 ACT and SAT data are provided in the README. These data live in PDFs, and so you'll get to enjoy practicing some *manual* data collection. Save these data as a CSV in your `data` directory, and import, explore, and clean these data in the same way you did above. **Make sure you comment on your steps so it is clear *why* you are doing each process**.

In [270]:
sat_2018 = pd.read_csv("../data/SAT2018_Testing _Data - SAT 2018.csv")
act_2018 = pd.read_csv('../data/SAT2018_Testing _Data - ACT 2018.csv')

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

In [271]:
# make sure the data type
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 3 columns):
State            52 non-null object
Participation    52 non-null object
Composite        52 non-null float64
dtypes: float64(1), object(2)
memory usage: 1.3+ KB


In [272]:
# covert the participation to float for both sat_2018 & act_2018
sat_2018['Participation'] = sat_2018['Participation'].map(lambda cell: float(cell.replace("%", ""))/100)
print(sat_2018.info())
act_2018['Participation'] = act_2018['Participation'].map(lambda cell: float(cell.replace("%",""))/100)
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null float64
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 3 columns):
State            52 non-null object
Participation    52 non-null float64
Composite        52 non-null float64
dtypes: float64(2), object(1)
memory usage: 1.3+ KB


In [273]:
# convert sat_2018 from int to float
sat_2018_new = as_float(sat_2018, 'Evidence-Based Reading and Writing' )
sat_2018_new = as_float(sat_2018, 'Math' )
sat_2018_new = as_float(sat_2018, 'Total' )

In [274]:
# change the column_name for both sat and act
sat_2018_new_col = {'State':'state' ,
                     'Participation':'participation_sat18',
                     'Evidence-Based Reading and Writing':'reading_writing_sat18',
                     'Math':'math_sat18',
                     'Total':'total_sat18'} 
sat_2018_new.rename(columns = sat_2018_new_col, inplace = True)
print(sat_2018_new.head())
act_2018_new = act_2018
act_2018_new_col = {'State':'state' ,
                     'Participation':'participation_act18',
                     'Composite':'total_act18'} 
act_2018_new.rename(columns = act_2018_new_col, inplace = True)
act_2018_new.head()


          state  participation_sat18  reading_writing_sat18  math_sat18  \
0     Minnesota                 0.04                  643.0       655.0   
1     Wisconsin                 0.03                  641.0       653.0   
2  North Dakota                 0.02                  640.0       643.0   
3          Iowa                 0.03                  634.0       631.0   
4        Kansas                 0.04                  633.0       631.0   

   total_sat18  
0       1298.0  
1       1294.0  
2       1283.0  
3       1265.0  
4       1265.0  


state  participation_act18  total_act18
0     Alabama                 1.00         19.1
1      Alaska                 0.33         20.8
2     Arizona                 0.66         19.2
3    Arkansas                 1.00         19.4
4  California                 0.27         22.7

In [277]:
# merge the sat_2018 & act_2018 together
combined_2018 = pd.DataFrame.merge(sat_2018_new, right = act_2018_new, how ='inner')
combined_2018.head()

state  participation_sat18  reading_writing_sat18  math_sat18  \
0     Minnesota                 0.04                  643.0       655.0   
1     Wisconsin                 0.03                  641.0       653.0   
2  North Dakota                 0.02                  640.0       643.0   
3          Iowa                 0.03                  634.0       631.0   
4        Kansas                 0.04                  633.0       631.0   

   total_sat18  participation_act18  total_act18  
0       1298.0                 0.99         21.3  
1       1294.0                 1.00         20.5  
2       1283.0                 0.98         20.3  
3       1265.0                 0.68         21.8  
4       1265.0                 0.71         21.6

In [279]:
# merge the combine_2017 with combined_2018
final_all = pd.DataFrame.merge(combined_2017, right = combined_2018, how = 'inner')
final_all.head()

state  participation_sat17  reading_writing_sat17  math_sat17  \
0     Alabama                 0.05                  593.0       572.0   
1      Alaska                 0.38                  547.0       533.0   
2     Arizona                 0.30                  563.0       553.0   
3    Arkansas                 0.03                  614.0       594.0   
4  California                 0.53                  531.0       524.0   

   total_sat17  participation_act17  english_act17  math_act17  reading_act17  \
0       1165.0                 1.00           18.9        18.4           19.7   
1       1080.0                 0.65           18.7        19.8           20.4   
2       1116.0                 0.62           18.6        19.8           20.1   
3       1208.0                 1.00           18.9        19.0           19.7   
4       1055.0                 0.31           22.5        22.7           23.1   

   science_act17  total_act17  participation_sat18  reading_writing_sat18  \
0           19.4         19.2                 0.06                  595.0   
1           19.9         19.8                 0.43                  562.0   
2           19.8         19.7                 0.29                  577.0   
3           19.5         19.4                 0.05                  592.0   
4           22.2         22.8                 0.60                  540.0   

   math_sat18  total_sat18  participation_act18  total_act18  
0       571.0       1166.0                 1.00         19.1  
1       544.0       1106.0                 0.33         20.8  
2       572.0       1149.0                 0.66         19.2  
3       576.0       1169.0                 1.00         19.4  
4       536.0       1076.0                 0.27         22.7

In [280]:
# save the clean final.csv file
final_all.to_csv('final_all.csv')

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [ ]:
#Code:

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [ ]:
#code

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [ ]:
#Code:

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

Answer

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [ ]:
#code

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [ ]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels):
    nrows = int(np.ceil(len(list_of_columns)/2) # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        ax[i].hist(dataframe[column]) # feel free to add more settings
        # Set titles, labels, etc here for each subplot

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT
- Math scores for SAT & ACT
- Reading/verbal scores for SAT & ACT

In [ ]:
# Code

#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
# Code

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [ ]:
# Code

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

Answers:

#### We generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?

Answer:

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer:

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

Answer:

##### Is it appropriate to compare *these* specific SAT and ACT math scores? 

Why or why not?

Answer:

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code:

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?

**Thank you for the help from local instructor Boom and Laura and the GA Global Lectures instructor. **

### Online Resources:
   - https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html
   - https://stackoverflow.com/questions/32400867/pandas-read-csv-from-url
   - https://www.w3schools.com/python/python_lambda.asp
   - https://www.geeksforgeeks.org/difference-between-map-applymap-and-apply-methods-in-pandas/
   - https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html
   - https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html
   - https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html
   - https://www.geeksforgeeks.org/saving-a-pandas-dataframe-as-a-csv/
   - 
   - 
   - 